File: Lab 1/board.py
EMPTY = " "
MARK = "x"

DIRECTIONS = [(dx, dy) for dx in range(-1, 2) for dy in range(-1, 2) if dx or dy]

class Board:
    def __init__(self, black_symbol, white_symbol):
        self.state = [[EMPTY] * 10 for _ in range(10)]
        self.state[4][5] = self.state[5][4] = black_symbol
        self.state[4][4] = self.state[5][5] = white_symbol
        self.valid_positions = [(x, y) for x in range(1, 9) for y in range(1, 9) if self.state[x][y] == EMPTY]
    
    def __str__(self):
        top = "    " + "   ".join("ABCDEFGH")
        line = "-" * 33
        rows = f"\n  {line}\n".join(f"{i} | {' | '.join(x[1:-1])} |" for i, x in enumerate(self.state[1:-1], 1))
        return f"{top}\n  {line}\n{rows}\n  {line}"
    
    def stringify_with_moves(self, moves):
        moves = [(int(x), ord(y) - 64) for y, x in moves]
        for x, y in moves:
            self.state[x][y] = MARK
        result = str(self)
        for x, y in moves:
            self.state[x][y] = EMPTY
        return result

    def find_moves(self, player_symbol):
        moves = []
        for x, y in self.valid_positions:
            for dx, dy in DIRECTIONS:
                if self.state[x][y] == EMPTY and self.line_exists(player_symbol, x, y, dx, dy):
                    moves.append(f"{chr(y + 64)}{x}")
                    break
        return moves
    
    def line_exists(self, player_symbol, x, y, dx, dy):
        X = x
        Y = y
        while True:
            X += dx
            Y += dy
            if self.state[X][Y] == EMPTY:
                return False
            if self.state[X][Y] == player_symbol:
                return X - x != dx or Y - y != dy
    
    def make_move(self, move, player_symbol):
        x = int(move[1])
        y = ord(move[0]) - 64
        self.state[x][y] = player_symbol
        for dx, dy in DIRECTIONS:
            if self.line_exists(player_symbol, x, y, dx, dy):
                X = x
                Y = y
                while True:
                    X += dx
                    Y += dy
                    if self.state[X][Y] == player_symbol:
                        break
                    self.state[X][Y] = player_symbol
    
    def calculate_total(self, black_symbol, white_symbol):
        black_total = sum(x.count(black_symbol) for x in self.state)
        white_total = sum(x.count(white_symbol) for x in self.state)
        return (black_total, white_total)

File: Lab 2/Board.cs
using System.Collections.Generic;

namespace Antireversi {
    internal class Board {
        private static readonly List<(int, int)> Directions = new List<(int, int)> {
            (-1, -1), (-1, 0), (-1, 1),
            ( 0, -1), ( 0, 1),
            ( 1, -1), ( 1, 0), ( 1, 1)
        };

        public readonly CellValue[,] State;
        public readonly List<(int, int)> ValidPositions;

        public Board() {
            State = new CellValue[10, 10];
            for (var i = 0; i < 10; i++)
                for (var j = 0; j < 10; j++)
                    State[i, j] = CellValue.Empty;
            
            State[4, 5] = CellValue.Black;
            State[5, 4] = CellValue.Black;
            
            State[4, 4] = CellValue.White;
            State[5, 5] = CellValue.White;
            
            ValidPositions = new List<(int, int)>();
            for (var i = 1; i <= 8; i++)
                for (var j = 1; j <= 8; j++)
                    if (State[i, j] == CellValue.Empty)
                        ValidPositions.Add((i, j));
        }

        public Board(Board board, (int, int) move, CellValue c) {
            State = new CellValue[10, 10];
            for (var i = 0; i < 10; i++)
                for (var j = 0; j < 10; j++)
                    State[i, j] = board.State[i, j];
            
            ValidPositions = new List<(int, int)>(board.ValidPositions);
            
            MakeMove(move, c);
        }

        public static string EncodeMove((int, int) move) {
            var (x, y) = move;
            return $"{(char) (y + 64)}{x}";
        }

        public static (int, int) DecodeMove(string move) {
            var x = move[1] - '0';
            var y = move[0] - 64;
            return (x, y);
        }

        public List<(int, int)> FindMoves(CellValue c) {
            var moves = new List<(int, int)>();
            
            foreach (var (x, y) in ValidPositions)
                foreach (var (dx, dy) in Directions)
                    if (LineExists(c, x, y, dx, dy)) {
                        moves.Add((x, y));
                        break;
                    }
            
            return moves;
        }

        private bool LineExists(CellValue c, int x, int y, int dx, int dy) {
            var xx = x;
            var yy = y;
            
            while (true) {
                xx += dx;
                yy += dy;
                
                if (State[xx, yy] == CellValue.Empty || State[xx, yy] == CellValue.Blackhole)
                    return false;
                if (State[xx, yy] == c)
                    return xx - x != dx || yy - y != dy;
            }
        }

        public void MakeMove((int, int) move, CellValue c) {
            var (x, y) = move;
            State[x, y] = c;
            ValidPositions.Remove(move);
            
            foreach (var (dx, dy) in Directions)
                if (LineExists(c, x, y, dx, dy)) {
                    var xx = x;
                    var yy = y;
                    
                    while (true) {
                        xx += dx;
                        yy += dy;
                        
                        if (State[xx, yy] == c)
                            break;
                        
                        State[xx, yy] = c;
                    }
                }
        }

        public int Count(CellValue c) {
            var result = 0;
            
            for (var i = 1; i <= 8; i++)
                for (var j = 1; j <= 8; j++)
                    if (State[i, j] == c)
                        result++;
            
            return result;
        }

        public void AddBlackHole((int, int) pos) {
            var (x, y) = pos;
            State[x, y] = CellValue.Blackhole;
            ValidPositions.Remove(pos);
        }
    }
}